### PR TITLE
Account for rounding errors when deciding the diagnostic boundaries

### LINKED ISSUE #64020: `miri` no longer builds after rust-lang/rust#63402
Hello, this is your friendly neighborhood mergebot.
After merging PR rust-lang/rust#63402, I observed that the tool miri has failing tests.
A follow-up PR to the repository https://github.com/rust-lang/miri is needed to fix the fallout.

cc @estebank, do you think you would have time to do the follow-up work?
If so, that would be great!

cc @oli-obk, the PR reviewer, and @rust-lang/compiler -- nominating for prioritization.

### PR DESCRIPTION
**Pull request by estebank**
Fix Miri by fixing the bug raised in https://github.com/rust-lang/rust/pull/63402#discussion_r319485328.

Fixes https://github.com/rust-lang/rust/issues/64020

### CODE REVIEW

**Comment by estebank** [comment]
Screen Shot 2019-08-30 at 5 47 08 PM
Screen Shot 2019-08-30 at 5 46 53 PM

CC @RalfJung

File: src/librustc_errors/emitter.rs [hunk 1]
```diff
@@ -146,12 +146,12 @@ impl Margin {
             } else if self.label_right - self.span_left <= self.column_width {
                 // Attempt to fit the code window considering only the spans and labels.
                 let padding_left = (self.column_width - (self.label_right - self.span_left)) / 2;
-                self.computed_left = self.span_left - padding_left;
+                self.computed_left = max(self.span_left, padding_left) - padding_left;
```
Comment (RalfJung) [inline src/librustc_errors/emitter.rs]: These could also all use `saturating_sub` instead, right? What do you think would be more readable?

**Comment by RalfJung** [comment]
Also I wonder, is there any way you could add a test for this? (But IMO that's not a blocker.)

**Comment by RalfJung** [comment]
I can confirm that this fixes Miri. r=me with the comment about `saturating_sub` resolved (one way or another).

**Comment by estebank** [comment]
> Also I wonder, is there any way you could add a test for this?

I haven't found why the `compile-flags` doesn't propagate `-Zterminal-width` properly to tests. I'll get back to it to increase test coverage, but for now I'm letting it slide with some manual testing 😬

File: src/librustc_errors/emitter.rs [hunk 2]
```diff
@@ -1304,11 +1304,13 @@ impl EmitterWriter {
...
                 let column_width = if let Some(width) = self.terminal_width {
-                    width
+                    width.saturating_sub(code_offset)
                 } else if self.ui_testing {
                     140
                 } else {
-                    term_size::dimensions().map(|(w, _)| w - code_offset).unwrap_or(140)
+                    term_size::dimensions()
+                        .map(|(w, _)| w.saturating_sub(code_offset))
+                        .unwrap_or(std::usize::MAX)
```
Comment (bjorn3) [inline src/librustc_errors/emitter.rs]: Why was this changed to usize::MAX?

Comment (estebank) [reply src/librustc_errors/emitter.rs]: @bjorn3 If the dimensions aren't supplied and can't be queried, I decided that keeping the _current_ uncropped behavior makes more sense. By setting it to `MAX` it'll never crop.

**Pull request merged**